src/syscalls_new.rs

```diff
@@ -0,0 +1,190 @@
+use crate::callback::CallbackSubscription;
+use crate::callback::SubscribableCallback;
+use crate::shared_memory::SharedMemory;
+use core::isize;
+
+#[derive(Debug)]
+pub enum Error {
+    Fail,           // Generic failure condition
+    Busy,           // Underlying system is busy; retry
+    Already,        // The state requested is already set
+    Off,            // The component is powered down
+    Reserve,        // Reservation required before use
+    Invalid,        // An invalid parameter was passed
+    Size,           // Parameter passed was too large
+    Cancel,         // Operation cancelled by a call
+    Nomem,          // Memory required not available
+    Nosupport,      // Operation or command is unsupported
+    Nodevice,       // Device does not exist
+    Uninstalled,    // Device is not physically installed
+    Noack,          // Packet transmission not acknowledged
+    Unknown(isize), // Unknown Error
+}
+
+struct ReturnCode(isize);
+
+impl From<isize> for ReturnCode {
+    fn from(v: isize) -> ReturnCode {
+        ReturnCode(v)
+    }
+}
+
+impl From<ReturnCode> for Error {
+    fn from(v: ReturnCode) -> Error {
+        match v.0 {
+            -1 => Error::Fail,         // Generic failure condition
+            -2 => Error::Busy,         // Underlying system is busy; retry
+            -3 => Error::Already,      // The state requested is already set
+            -4 => Error::Off,          // The component is powered down
+            -5 => Error::Reserve,      // Reservation required before use
+            -6 => Error::Invalid,      // An invalid parameter was passed
+            -7 => Error::Size,         // Parameter passed was too large
+            -8 => Error::Cancel,       // Operation cancelled by a call
+            -9 => Error::Nomem,        // Memory required not available
+            -10 => Error::Nosupport,   // Operation or command is unsupported
+            -11 => Error::Nodevice,    // Device does not exist
+            -12 => Error::Uninstalled, //Device is not physically installed
+            -13 => Error::Noack,       // Packet transmission not acknowledged
+            _ => unreachable!(),
```

Comment:
This case is reachable for two reasons:
- A driver might return a value lower than -13. In that case the caller has no chance of telling what happened.
- Error::from(ReturnCode(n)) can be called for every possible n at the moment.

Reply:
Counter proposal: Rename `ReturnCode` to `ErrorCode` and implement `fn describe(&self) -> Option<Error>` on `ErrorType`.